File: frontend/app/src/services/admin.service.js
import API from "../api";

const getAuthHeader = () => {
    const user = JSON.parse(sessionStorage.getItem("user"));
    if (user && user.token) {
        return { Authorization: `Bearer ${user.token}` };
    } else {
        return {};
    }
};

const AdminService = {
    // --- Students ---
    getAllStudents: async () => {
        return await API.get("/admin/students");
    },
    getStudentById: async (id) => {
        return await API.get(`/admin/students/${id}`);
    },
    addStudent: async (studentData) => {
        return await API.post("/admin/students", studentData);
    },
    updateStudent: async (id, studentData) => {
        return await API.put(`/admin/students/${id}`, studentData);
    },
    deleteStudent: async (id) => {
        return await API.delete(`/admin/students/${id}`);
    },

    // --- Teachers ---
    getAllTeachers: async () => {
        return await API.get("/admin/teachers");
    },
    getTeacherById: async (id) => {
        return await API.get(`/admin/teachers/${id}`);
    },
    addTeacher: async (teacherData) => {
        return await API.post("/admin/teachers", teacherData);
    },
    updateTeacher: async (id, teacherData) => {
        return await API.put(`/admin/teachers/${id}`, teacherData);
    },
    deleteTeacher: async (id) => {
        return await API.delete(`/admin/teachers/${id}`);
    },

    // --- Courses ---
    getAllCourses: async () => {
        return await API.get("/admin/courses");
    },
    getCourseById: async (id) => {
        return await API.get(`/admin/courses/${id}`);
    },
    addCourse: async (courseData) => {
        return await API.post("/admin/courses", courseData);
    },
    updateCourse: async (id, courseData) => {
        return await API.put(`/admin/courses/${id}`, courseData);
    },
    deleteCourse: async (id) => {
        return await API.delete(`/admin/courses/${id}`);
    },

    // --- Subjects ---
    getAllSubjects: async () => {
        return await API.get("/admin/subjects");
    },
    getSubjectById: async (id) => {
        return await API.get(`/admin/subjects/${id}`);
    },
    addSubject: async (subjectData) => {
        return await API.post("/admin/subjects", subjectData, {
            headers: { "Content-Type": "multipart/form-data" }
        });
    },
    updateSubject: async (id, subjectData) => {
        // Check if subjectData is FormData (for file upload) or JSON
        if (subjectData instanceof FormData) {
            return await API.put(`/admin/subjects/${id}`, subjectData, {
                headers: { "Content-Type": "multipart/form-data" }
            });
        }
        return await API.put(`/admin/subjects/${id}`, subjectData);
    },
    deleteSubject: async (id) => {
        return await API.delete(`/admin/subjects/${id}`);
    },

    // --- Notices ---
    getAllNotices: async () => {
        return await API.get("/admin/notices");
    },
    addNotice: async (noticeData) => {
        return await API.post("/admin/notices", noticeData);
    },
    deleteNotice: async (id) => {
        return await API.delete(`/admin/notices/${id}`);
    },

    // --- Fee Management ---
    getAllStudentFees: async () => {
        return await API.get("/admin/fees");
    },
    updateFeeStatus: async (id, status) => {
        return await API.put(`/admin/fees/${id}/status`, status, {
            headers: { "Content-Type": "application/json" }
        });
    },

    // --- Student Marks ---
    getMarksByCourseAndSubject: async (courseId, subjectId) => {
        return await API.get(`/admin/marks/search?courseId=${courseId}&subjectId=${subjectId}`);
    },

    // --- Feedback Management ---
    getAllFeedbacks: async () => {
        return await API.get("/admin/feedbacks");
    },
    respondToFeedback: async (id, response) => {
        return await API.put(`/admin/feedbacks/${id}/respond`, response, {
            headers: { "Content-Type": "application/json" }
        });
    },

    // --- Attendance Reports ---
    getStudentAttendanceList: async () => {
        return await API.get("/admin/attendance/students");
    },
    getTeacherAttendanceList: async () => {
        return await API.get("/admin/attendance/teachers");
    },

    // --- User Status Management ---
    toggleUserStatus: async (id) => {
        return await API.put(`/admin/users/${id}/status`);
    },
    toggleCourseStatus: async (id) => {
        return await API.put(`/admin/courses/${id}/status`);
    },
    toggleSubjectStatus: async (id) => {
        return await API.put(`/admin/subjects/${id}/status`);
    },

    // --- Bulk Upload ---
    uploadUsers: async (file) => {
        const formData = new FormData();
        formData.append("file", file);
        return await API.post("/admin/users/upload", formData, {
            headers: {
                "Content-Type": "multipart/form-data",
            },
        });
    }
};

export default AdminService;

File: Backend/Institute_Management_System/src/main/java/com/ims/service/AdminService.java
package com.ims.service;

import com.ims.dto.CourseDTO;
import com.ims.dto.StudentDTO;
import com.ims.dto.SubjectDTO;
import com.ims.dto.AdminFeeDTO;
import com.ims.dto.AdminStudentMarksDTO;
import com.ims.dto.AdminFeedbackDTO;
import com.ims.dto.AdminStudentAttendanceDTO;
import com.ims.dto.AdminTeacherAttendanceDTO;
import com.ims.entity.User;
import com.ims.entity.Course;
import com.ims.entity.Subject;
import org.springframework.web.multipart.MultipartFile;
import java.util.List;

public interface AdminService {

    // Student Operations
    User addStudent(StudentDTO studentDTO);

    User updateStudent(Long id, StudentDTO studentDTO);

    void deleteStudent(Long id);

    List<User> getAllStudents();

    User getStudentById(Long id);

    // Teacher Operations
    User addTeacher(StudentDTO teacherDTO);

    User updateTeacher(Long id, StudentDTO teacherDTO);

    void deleteTeacher(Long id);

    List<User> getAllTeachers();

    User getTeacherById(Long id);

    // Course Operations
    Course addCourse(CourseDTO courseDTO);

    Course updateCourse(Long id, CourseDTO courseDTO);

    void deleteCourse(Long id);

    List<Course> getAllCourses();

    Course getCourseById(Long id);

    // Subject Operations
    Subject addSubject(SubjectDTO subjectDTO, org.springframework.web.multipart.MultipartFile scheduleFile);

    Subject updateSubject(Long id, SubjectDTO subjectDTO, MultipartFile scheduleFile);

    void deleteSubject(Long id);

    List<Subject> getAllSubjects();

    List<Subject> getSubjectsByCourseId(Long courseId);

    Subject getSubjectById(Long id);

    // Marks Operations
    List<AdminStudentMarksDTO> getMarksByCourseAndSubject(Long courseId, Long subjectId);

    // Fee Operations
    List<AdminFeeDTO> getAllStudentFees();

    AdminFeeDTO updateFeeStatus(Long feeId, String status);

    // Feedback Operations
    List<AdminFeedbackDTO> getAllFeedbacks();

    AdminFeedbackDTO respondToFeedback(Long feedbackId, String response);

    // Attendance Operations
    List<AdminStudentAttendanceDTO> getStudentAttendanceList();

    List<AdminTeacherAttendanceDTO> getTeacherAttendanceList();

    // Notice Operations
    com.ims.entity.Notice addNotice(com.ims.entity.Notice notice);

    List<com.ims.entity.Notice> getAllNotices();

    void deleteNotice(Long id);

    User toggleUserStatus(Long id);

    Course toggleCourseStatus(Long id);

    Subject toggleSubjectStatus(Long id);

    int saveUsers(MultipartFile file);
}
